### ISSUE TITLE
Recovering the color scale on bytescaled HV jp2 images.

### ISSUE THREAD

**Issue by alasdairwilson** [sunpy/sunpy#6715]
### Provide a general description of the issue or problem.

Related to #6710 ,

All jp2k files from HV are bytescaled in uint8, this means the array values when loaded into `Map` have a min and max of 0 and 255.

The actual physical unit colour limits of these images are hardcoded in the hv jp2gen files [e.g. for hmi](https://github.com/Helioviewer-Project/jp2gen/blob/master/idl/sdo/hmi/hvs_default_hmi.pro#L89) has a min of 0=-250 and 255 = +250.

However, as far as I can tell they don't write this to the jp2 file anywhere. The jp2 header has dataMin and dataMax fields but they are simply the fields taken straight from the origin fits file and not the min and max (and norm) used by the bytescaling process which is rarely linear anyway.

In principle JHelioviewer has keywords to handle this HV_ZERO and HV_SCALE which define the 0 on the color scale and the increment respectively and this works great. But this assumes, both in JHV and in a simple imagined solution for `Map` where just vmin and vmax are set, that the scaling is linear. Either way, since the keywords are not ever set by  HV as far as I can tell (by looking at common data sources' .jp2 files) then the question is fairly moot.

I don't have a good solution for this as, even though HMI is linearly scaled and so two numbers are all we need to define a scale (unlike most AIA channels); without these header keywords being present it relies on us hardcoding a color scale in which needs to match that defined in the HV code. This is probably no better than just leaving it to scale from 0 to 255 as it does currently by default.

edit: Also I guess since the interval is clipped then maybe it doesn't make sense to pretend these are real units anyway since a value of 255 in HMI could be a magnetic field of anything between 250 and 6000 gauss. To pretend the scale corresponds to a genuine physical unit (or datanumber) is probably more misleading than having it be arbitrary like it is now.

**Comment by alasdairwilson** [comment]
AS discussed in the dev meeting 18/01 won't fix as this is not science data, the clipped intervals make any conversion to real units meaningless.

**Comment by ayshih** [comment]
I don't agree with closing this issue.  Clipped data is still data.  We don't throw away EUV images simply because flare pixels saturate and thus are not accurate scientifically.  Not only are the non-clipped pixels still usable, it's also useful to know the clipping threshold in physical units.

For the non-clipped pixels, ideally the information would be carried in the JP2 header.  Since it apparently isn't, I propose creating an issue for `hvpy` to provide a way to query their scaling parameters for each source.  It'd be their responsibility to accurately report what they actually do (and keep synchronized to any changes).  In the meantime, we can go ahead with hard-coded parameters; this wouldn't be all that different from the hard-coding we do elsewhere.

For the clipped pixels, we can use masking.  Or we can use a default vmin/vmax of 1 to 254 so that we can leverage matplotlib's handling of out-of-range pixels.  Or we could issue a warning if we're really concerned that a user won't appreciate that the data is clipped.  Or we could do nothing, as we currently do for all map sources, even when they may have saturated pixels.

I won't reopen this issue myself, as I wasn't in the dev telecon, so I don't know whether @alasdairwilson's sentence fully captures the discussion.  But, I'd like this issue to be given more consideration.

**Comment by alasdairwilson** [comment]
I'll reopen it if there is a potential for useful discussion. The thinking was that the images usefulness should only be as pictures of the sun rather than Data and without the scaling information being in the header then the only way to recover the scaling is to manually apply it on a per channel/instr basis.

I appreciate the point about the instrument saturation but DN is what the instrument measured even if it were saturated, the recovered values from these clipped scales will not represent the values read from the instrument. The clipping is actually quite extreme as well, much of the dynamic range is removed.

We could, like you say use 1 to 254 and we could grab the following 3 bits of info from hv source code: dataMin, dataMax and scaling. That would be a pathway to reconstructing the colorscale.

**Comment by wafels** [comment]
The clipping/scaling information for at least some of the Helioviewer JPEG2000 files is already available in the J[P2Gen repository](https://github.com/Helioviewer-Project/jp2gen) but it is a little difficult to know where to get it. Not all institutions that provide Helioviewer JPEG2000 files use JP2Gen, and it is possible that institutions that use JP2Gen may have tweaked the processing pipeline without passing changes back to JP2Gen. See also https://github.com/Helioviewer-Project/api/issues/111 . The clipping/scaling information is also only useful for understanding how a prepped FITS file is transformed into a byte-scaled image. Inverting the JPEG2000 clipping/scaling process will result in an approximation to the original prepped FITS file.
